### PR TITLE
Instance association using IoU of 3d bounding boxes

### PR DESCRIPTION
**Pull request by yvsriram**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py [hunk 1]
```diff
@@ -171,21 +182,91 @@ def reset(self):
...
+            np.prod(global_bboxes_max - global_bboxes_min, axis=-1)
+            + np.prod(local_bbox[1] - local_bbox[0])
+            - intersection
+        )
+        ious = intersection / union
+        ious[zero_iou] = 0.0
+        ious[np.isnan(ious)] = 0.0
+        return ious
+
+    def find_global_instance_by_bbox_overlap(self, env_id: int, local_instance_id: int):
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:211]: Could we add some docstring describing these functions?

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py [hunk 2]
```diff
@@ -171,21 +182,91 @@ def reset(self):
...
         # get instance view
+        instance_view = self.get_local_instance_view(env_id, local_instance_id)
+        if instance_view is not None:
+            ious = self.get_bbox_overlap(instance_view.bounds, global_bounds)
+            if ious.max() > self.iou_threshold:
+                return global_instance_ids[ious.argmax()], ious.max()
+        return None, None
+
+    def associate_instances_to_memory(self):
+        # for each instance view, find the best matching instance
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:229]: formatting as docstring

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py [hunk 3]
```diff
@@ -154,10 +156,19 @@ class InstanceMemory:
     unprocessed_views: List[Dict[int, InstanceView]] = []
     timesteps: List[int] = []
 
-    def __init__(self, num_envs: int, du_scale: int, debug_visualize: bool = False):
+    def __init__(
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:159]: can we change the instance_map, segmentation_map names to something more generic + add a docstring? also make segmentation optional

**Review by cpaxton — APPROVED** [review]
lgtm with some minor documentation updates

**Comment by alexsax** [comment]
lgtm too!

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py (hunk 3 above)
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:101]: Changed to instance_seg, semantic_seg. Sorry for the confusion.

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py (hunk 2 above)
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py]: Done

**Pull request merged**